generation.py

```diff
@@ -40,45 +36,44 @@ def generate(self, models, args):
             if self.env.terminal():
                 break
 
+            moment = {'observation': {}, 'value': {}}
+
             for index, player in enumerate(self.env.players()):
-                observation, v = None, None
+                obs, v = None, None
                 if player == self.env.turn() or self.args['observation']:
-                    observation = self.env.observation(player)
+                    obs = self.env.observation(player)
                     model = models[player]
-                    p, v, hidden[player] = model.inference(observation, hidden[player])
+                    p, v, hidden[player] = model.inference(obs, hidden[player])
                     if player == self.env.turn():
                         legal_actions = self.env.legal_actions()
                         pmask = np.ones_like(p) * 1e32
                         pmask[legal_actions] = 0
                         p_turn = p - pmask
                         index_turn = index
-                observations[index].append(observation)
-                values[index].append(v)
+                moment['observation'][index] = obs
+                moment['value'][index] = v
 
             action = random.choices(legal_actions, weights=softmax(p_turn[legal_actions]))[0]
 
-            policies.append(p_turn)
-            pmasks.append(pmask)
-            turns.append(index_turn)
-            behaviors.append(action)
+            moment['policy'] = p_turn
+            moment['pmask'] = pmask
+            moment['turn'] = index_turn
+            moment['action'] = action
+            moments.append(bz2.compress(pickle.dumps(moment)))
```

Comment:
One suggestion. I guess it might be good to try multiple methods of dump/load (like joblib) and compress/decompress (like gzip) for efficient implementation.

- Ref
  - http://gael-varoquaux.info/programming/new_low-overhead_persistence_in_joblib_for_big_data.html
  - https://www.rootusers.com/gzip-vs-bzip2-vs-xz-performance-comparison/